pydicom/filewriter.py

```diff
@@ -431,15 +423,15 @@ def write_data_element(fp, data_element, encoding=default_encoding):
                     not val.startswith(b'\xff\xfe\xe0\x00')):
                 raise ValueError('Pixel Data with undefined length must '
                                  'start with an item tag')
+
     location = buffer.tell()
-    buffer.seek(0)
     if not fp.is_implicit_VR and VR not in extra_length_VRs:
-        buffer.write_US(location - 2)  # 2 is length of US
+        fp.write_US(location)  # Explicit VR length field is only 2 bytes
     else:
-        # write the proper length of the data_element back in the length slot,
+        # write the proper length of the data_element in the length slot,
         # unless is SQ with undefined length.
-        if not is_undefined_length:
-            buffer.write_UL(location - 4)  # 4 is length of UL
+        fp.write_UL(0xFFFFFFFF if is_undefined_length else location)
```

Comment:
the base `location` on end looks wrong here - shouldn't it be a write?

Reply:
Not sure if I understand - location is the buffer size (the length of the tag -actually I should rename it), and is written before the buffer content is written.

Reply:
Never mind, my mistake, I missed that it is within the brackets.

Reply:
Ok, but good you asked - I renamed `location` to `tag_length` now.

Reply:
I like the rename, its easier to understand, but ... if being really picky on DICOM terminology, isn't it the value length?

Reply:
Yes, you are right, of course. I will rename it properly.

Reply:
Done.